### PR TITLE
Switching SecureTokens to Base58

### PR DESCRIPTION
**Pull request by robertomiranda**
ref https://github.com/rails/rails/pull/18217#discussion_r22443992

### CODE REVIEW

**Comment by robertomiranda** [comment]
cc @dhh, @chancancode

**Comment by dhh** [comment]
Looks good to me, but I'll let someone in the know actually review the algorithm ;).

File: activesupport/lib/active_support/core_ext/securerandom.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  # The argument _n_ specifies the length, in bytes, of the random number
+  # to be generated.
+  #
+  # If _n_ is not specified or is nil, 16 is assumed.
+  # It may be larger in the future.
+  #
+  # The result may contain A-Z, a-z, 0-9
+  #
+  #   p SecureRandom.base64 #=> "08101M3q0p1v0z2h21171Y413e2M3r0t"
+  #   p SecureRandom.base64 #=> "3D3H0M3m1T3c1E2P290I0C0t3f432t1J"
```
Comment (vipulnsward) [inline activesupport/lib/active_support/core_ext/securerandom.rb]: Should be examples of `base62`

File: activesupport/test/core_ext/securerandom_test.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+require 'abstract_unit'
+require 'active_support/core_ext/securerandom'
+class SecureRandomTest < ActiveSupport::TestCase
+  def test_generate_base62_string
+    puts SecureRandom.base62
```
Comment (vipulnsward) [inline activesupport/test/core_ext/securerandom_test.rb]: `puts` is not needed

**Comment by matthewd** [comment]
An actual base62 encoder seems needlessly complicated for our purposes... see my comments on previous discussion.

**Comment by robertomiranda** [comment]
whats's of issue with use base64 instead of base62?

**Comment by matthewd** [comment]
@robertomiranda it contains characters we don't want. That was also discussed to death in the previous PR.

**Comment by dhh** [comment]
@matthewd What are you suggesting as an alternative here?

**Comment by matthewd** [comment]
Something like:

```ruby
random_bytes(chars).unpack("C*").map do |byte|
  idx = byte % 64
  idx = random_number(62) if idx >= 62
  BASE62_CHARS[idx]
end.join
```

And having looked closer, I've just realised that @robertomiranda's implementation is actually Wrong: it's not generating a uniformly random string.

File: activesupport/lib/active_support/core_ext/securerandom.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'active_support/base62'
+
+module SecureRandom
+  # SecureRandom.base62 generates a random base62 string.
+  #
+  # The argument _n_ specifies the length, in bytes, of the random number
```
Comment (georgeclaghorn) [inline activesupport/lib/active_support/core_ext/securerandom.rb]: The argument is named `bytes`, not `n`.

**Comment by matthewd** [comment]
The above is obviously assuming that `random_bytes` is faster / otherwise betterer than a series of calls to `random_number`... otherwise, life gets much simpler:

```ruby
''.tap do |string|
  length.times { string << BASE62_CHARS[random_number(62)] }
end
```

**Comment by egilburg** [comment]
>it's not generating a uniformly random string.

`idx = random_number(62) if idx >= 62` is not uniform either (it'll produce a slight bias towards integers than towards letters). But is the goal here to produce a cryptographically secure random string, or something "close enough"?

Also consider readability for those who may want to write these strings down - some existing algorithms, for exampe, avoid uppercase I (i) as it looks like lowercase l (L), same with o/O/0.

**Comment by chancancode** [comment]
> Also consider readability for those who may want to write these strings down - some existing algorithms, for exampe, avoid uppercase I (i) as it looks like lowercase l (L), same with o/O/0.

Assuming a uniform distribution, with [Base58](http://en.wikipedia.org/wiki/Base58) at 24 characters we are still getting 140 random bits, which is still miles better than a random UUID, so I guess why not. It's useful for things like "write down these TFA lock out code in case you lost your phone", or "enter this manually if your phone can't scan this QR code" scenario.

> But is the goal here to produce a cryptographically secure random string, or something "close enough"?

IMO, "close enough" is a dangerous territory here. Once you are off the tracks, all bets are off and it's difficult to tell what you are up against (e.g. Is this still equivalent or better than a random UUID? How would we prove it?). It also makes it difficult to put any guarantee on it hence recommend using this API for any specific purpose.

**Comment by dhh** [comment]
:+1: on what Godfrey said.

**Comment by matthewd** [comment]
@egilburg I'm obviously being thick, because that looks uniform to me. Care to expand?

**Comment by egilburg** [comment]
@matthewd perhaps I'm the thicker one here, but this is a bit confusing to me:

```ruby
  idx = byte % 64
  idx = random_number(62) if idx >= 62
```

It seems you get one of 64 variations, but 2 of those are "bad" so you "reroll" them to a backup implementation. Why would either just `byte % 62` or just do `idx = random_number(62)` work then?

**Comment by matthewd** [comment]
256 values in a byte; `% 64` is safe (drops two bits), `% 62` would bias.

I did offer a vastly simpler alternative that just uses `random_number` from the outset... this version was building on an (untested) assumption that doing most of the work with a single call to `random_bytes` might be noticeably more efficient.

**Pull request merged**